File: mobile/lib/models/product.dart
class Product {
  final String id;
  final String name;
  final String batchNumber;
  final String farmerID;
  final Location farmLocation;
  final DateTime plantingDate;
  final DateTime harvestDate;
  final String quality;
  final List<String> certifications;
  final String currentOwner;
  final Location currentLocation;
  final ProductStatus status;
  final List<SupplyChainStep> supplyChainSteps;
  final String? ipfsHash;
  final String? qrCode;
  final DateTime createdAt;
  final DateTime updatedAt;

  Product({
    required this.id,
    required this.name,
    required this.batchNumber,
    required this.farmerID,
    required this.farmLocation,
    required this.plantingDate,
    required this.harvestDate,
    required this.quality,
    required this.certifications,
    required this.currentOwner,
    required this.currentLocation,
    required this.status,
    required this.supplyChainSteps,
    this.ipfsHash,
    this.qrCode,
    required this.createdAt,
    required this.updatedAt,
  });

  factory Product.fromJson(Map<String, dynamic> json) {
    return Product(
      id: json['id'] ?? '',
      name: json['name'] ?? '',
      batchNumber: json['batchNumber'] ?? '',
      farmerID: json['farmerID'] ?? '',
      farmLocation: Location.fromJson(json['farmLocation'] ?? {}),
      plantingDate: DateTime.parse(json['plantingDate']),
      harvestDate: DateTime.parse(json['harvestDate']),
      quality: json['quality'] ?? '',
      certifications: List<String>.from(json['certifications'] ?? []),
      currentOwner: json['currentOwner'] ?? '',
      currentLocation: Location.fromJson(json['currentLocation'] ?? {}),
      status: ProductStatusExtension.fromString(json['status'] ?? 'PLANTED'),
      supplyChainSteps: (json['supplyChainSteps'] as List<dynamic>?)
              ?.map((step) => SupplyChainStep.fromJson(step))
              .toList() ??
          [],
      ipfsHash: json['ipfsHash'],
      qrCode: json['qrCode'],
      createdAt: DateTime.parse(json['createdAt']),
      updatedAt: DateTime.parse(json['updatedAt']),
    );
  }

  Map<String, dynamic> toJson() {
    return {
      'id': id,
      'name': name,
      'batchNumber': batchNumber,
      'farmerID': farmerID,
      'farmLocation': farmLocation.toJson(),
      'plantingDate': plantingDate.toIso8601String(),
      'harvestDate': harvestDate.toIso8601String(),
      'quality': quality,
      'certifications': certifications,
      'currentOwner': currentOwner,
      'currentLocation': currentLocation.toJson(),
      'status': status.toString().split('.').last,
      'supplyChainSteps': supplyChainSteps.map((step) => step.toJson()).toList(),
      'ipfsHash': ipfsHash,
      'qrCode': qrCode,
      'createdAt': createdAt.toIso8601String(),
      'updatedAt': updatedAt.toIso8601String(),
    };
  }
}

class Location {
  final double latitude;
  final double longitude;
  final String address;

  Location({
    required this.latitude,
    required this.longitude,
    required this.address,
  });

  factory Location.fromJson(Map<String, dynamic> json) {
    return Location(
      latitude: (json['latitude'] ?? 0.0).toDouble(),
      longitude: (json['longitude'] ?? 0.0).toDouble(),
      address: json['address'] ?? '',
    );
  }

  Map<String, dynamic> toJson() {
    return {
      'latitude': latitude,
      'longitude': longitude,
      'address': address,
    };
  }
}

enum ProductStatus {
  planted,
  harvested,
  processed,
  inTransit,
  delivered,
  sold,
}

extension ProductStatusExtension on ProductStatus {
  String get displayName {
    switch (this) {
      case ProductStatus.planted:
        return 'Planted';
      case ProductStatus.harvested:
        return 'Harvested';
      case ProductStatus.processed:
        return 'Processed';
      case ProductStatus.inTransit:
        return 'In Transit';
      case ProductStatus.delivered:
        return 'Delivered';
      case ProductStatus.sold:
        return 'Sold';
    }
  }

  static ProductStatus fromString(String status) {
    switch (status.toUpperCase()) {
      case 'PLANTED':
        return ProductStatus.planted;
      case 'HARVESTED':
        return ProductStatus.harvested;
      case 'PROCESSED':
        return ProductStatus.processed;
      case 'IN_TRANSIT':
        return ProductStatus.inTransit;
      case 'DELIVERED':
        return ProductStatus.delivered;
      case 'SOLD':
        return ProductStatus.sold;
      default:
        return ProductStatus.planted;
    }
  }
}

class SupplyChainStep {
  final String id;
  final String stepType;
  final String actor;
  final Location location;
  final DateTime timestamp;
  final String description;
  final Map<String, String> metadata;

  SupplyChainStep({
    required this.id,
    required this.stepType,
    required this.actor,
    required this.location,
    required this.timestamp,
    required this.description,
    required this.metadata,
  });

  factory SupplyChainStep.fromJson(Map<String, dynamic> json) {
    return SupplyChainStep(
      id: json['id'] ?? '',
      stepType: json['stepType'] ?? '',
      actor: json['actor'] ?? '',
      location: Location.fromJson(json['location'] ?? {}),
      timestamp: DateTime.parse(json['timestamp']),
      description: json['description'] ?? '',
      metadata: Map<String, String>.from(json['metadata'] ?? {}),
    );
  }

  Map<String, dynamic> toJson() {
    return {
      'id': id,
      'stepType': stepType,
      'actor': actor,
      'location': location.toJson(),
      'timestamp': timestamp.toIso8601String(),
      'description': description,
      'metadata': metadata,
    };
  }
}

File: contracts/agri-chaincode/model/product.go
package model

import (
	"time"
)

// Product represents an agricultural product in the supply chain
type Product struct {
	ID                string            `json:"id"`
	Name              string            `json:"name"`
	BatchNumber       string            `json:"batchNumber"`
	FarmerID          string            `json:"farmerID"`
	FarmLocation      Location          `json:"farmLocation"`
	PlantingDate      time.Time         `json:"plantingDate"`
	HarvestDate       time.Time         `json:"harvestDate"`
	Quality           string            `json:"quality"`
	Certifications    []string          `json:"certifications"`
	CurrentOwner      string            `json:"currentOwner"`
	CurrentLocation   Location          `json:"currentLocation"`
	Status            ProductStatus     `json:"status"`
	SupplyChainSteps  []SupplyChainStep `json:"supplyChainSteps"`
	IPFSHash          string            `json:"ipfsHash"`
	QRCode            string            `json:"qrCode"`
	CreatedAt         time.Time         `json:"createdAt"`
	UpdatedAt         time.Time         `json:"updatedAt"`
}

// Location represents geographical coordinates
type Location struct {
	Latitude  float64 `json:"latitude"`
	Longitude float64 `json:"longitude"`
	Address   string  `json:"address"`
}

// ProductStatus represents the current status of the product
type ProductStatus string

const (
	StatusPlanted    ProductStatus = "PLANTED"
	StatusHarvested  ProductStatus = "HARVESTED"
	StatusProcessed  ProductStatus = "PROCESSED"
	StatusInTransit  ProductStatus = "IN_TRANSIT"
	StatusDelivered  ProductStatus = "DELIVERED"
	StatusSold       ProductStatus = "SOLD"
)

// SupplyChainStep represents a step in the supply chain
type SupplyChainStep struct {
	ID          string            `json:"id"`
	StepType    string            `json:"stepType"`
	Actor       string            `json:"actor"`
	Location    Location          `json:"location"`
	Timestamp   time.Time         `json:"timestamp"`
	Description string            `json:"description"`
	Metadata    map[string]string `json:"metadata"`
}

// Certificate represents quality/organic certifications
type Certificate struct {
	ID          string    `json:"id"`
	ProductID   string    `json:"productID"`
	Type        string    `json:"type"`
	IssuedBy    string    `json:"issuedBy"`
	IssuedDate  time.Time `json:"issuedDate"`
	ValidUntil  time.Time `json:"validUntil"`
	IPFSHash    string    `json:"ipfsHash"`
	Status      string    `json:"status"`
}

// Farmer represents a farmer in the system
type Farmer struct {
	ID            string            `json:"id"`
	Name          string            `json:"name"`
	Email         string            `json:"email"`
	Phone         string            `json:"phone"`
	FarmLocation  Location          `json:"farmLocation"`
	Certifications []string         `json:"certifications"`
	Verified      bool              `json:"verified"`
	Metadata      map[string]string `json:"metadata"`
	CreatedAt     time.Time         `json:"createdAt"`
}